### PR TITLE
Examples for contructing expressions

### PR DESCRIPTION
**Pull request by siddhartha-gadgil**
A few examples for constructing expressions with some description of the functions used. These include both direct constructions and smart constructions.

### CODE REVIEW

**Review by arthurpaulino — CHANGES_REQUESTED** [review]
@siddhartha-gadgil thank you very much for the PR! Just two things to be addressed:
* Please add the output of those `#eval` calls as a comment to make it easier for the reader
* Add yourself as one of the authors on the root `README.md` file

**Comment by siddhartha-gadgil** [comment]
Thanks @arthurpaulino - will make the changes

**Comment by siddhartha-gadgil** [comment]
@arthurpaulino I have made the changes. The only problem is that in some cases the output is very long. I see a couple of options:

* have the long output (though with clutter)
* remove some `eval` commands
* replace some `eval` commands with `check` (which will have much smaller output).

Do let me know which option you feel is best.

**Comment by arthurpaulino** [comment]
@siddhartha-gadgil Seeing those long outputs of `#eval` makes me question their didactic purpose. I think we can remove the ocurrences of `#eval sumExpr 2 3 /-Lean.Expr.app`

Another thing is that you changed the definition of `sumExpr`. One is pure and the other isn't. And their difference is drastic enough to deserve an explanation (and even different names).

The `#eval` on the non-pure `sumExpr 2 3` is short enough and we can leave it there

**Comment by siddhartha-gadgil** [comment]
The point about the names is a good one. As with my other code, I will use names like `sumExprM` consistently for monadic stuff.
I will also remove the longer evals.

**Comment by siddhartha-gadgil** [comment]
I have removed the longer output and named things to reflect whether they are pure or monadic.

File: lean/expressions.lean [hunk 1]
```diff
@@ -114,4 +113,78 @@ This is why the signature of the `bvar` case is `Nat → Expr` and not
...
+def one := mkApp (mkConst ``Nat.succ) z
+#eval one /- Lean.Expr.app
+  (Lean.Expr.const `Nat.succ [] (Expr.mkData 3403344051 (bi := Lean.BinderInfo.default)))
+  (Lean.Expr.const `Nat.zero [] (Expr.mkData 3114957063 (bi := Lean.BinderInfo.default)))
+  (Expr.mkData 3354277877 (approxDepth := 1) (bi := Lean.BinderInfo.default)) -/
+
+def natExpr: Nat → Expr 
+| 0 => z
+| n + 1 => mkApp (mkConst ``Nat.succ) (natExpr n)
+#eval natExpr 3
```
Comment (arthurpaulino) [inline lean/expressions.lean:171]:
```suggestion
```

File: lean/expressions.lean [hunk 2]
```diff
@@ -114,4 +113,78 @@ This is why the signature of the `bvar` case is `Nat → Expr` and not
...
+| n + 1 => mkApp (mkConst ``Nat.succ) (natExpr n)
+#eval natExpr 3 
+
+/-!
+There are many helpers that make defining function applications easier. In the following we use the variant `mkAppN` which allows application with multiple arguments.
+Note that the expression we get is not simplified. Simplification requires working with `MetaM`, so will be considered in the chapter on `MetaM`.
+-/
+def sumExpr : Nat → Nat → Expr 
+| n, m => mkAppN (mkConst ``Nat.add) #[natExpr n, natExpr m]
+#eval sumExpr 2 3
```
Comment (arthurpaulino) [inline lean/expressions.lean:179]:
```suggestion
```

**Comment by siddhartha-gadgil** [comment]
I hope I have made all the changes suggested. Thanks for the feedback.

**Pull request merged**